drivers: i3c: Support I3C driver for STM32.

ExaltZephyr opened the pull request:
This PR introduces support for the I3C driver on STM32, enabling functionality APIs for I3C controllers.

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+		return -EIO;
+	}
+
+	/* Satisfying I3C start timing min timing will statisfy the rest of the conditions */
+	if (i2c_bus_freq > 400000) {
+		/* Mixed bus with I2C FM+ device */
+		free = (uint8_t)ceil((STM32_I3C_TBUF_FMP_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	} else if (i2c_bus_freq > 0) {
+		/* Mixed bus with I2C FM device */
+		free = (uint8_t)ceil((STM32_I3C_TBUF_FM_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	} else {
+		/* Pure I3C bus */
+		free = (uint8_t)ceil((STM32_I3C_TCAS_MIN_NS * i3c_clock / 1e9 - 0.5) / 2);
+	}
```
I don't think this is best way to determine if it should be in pure, mixed fast, or mixed slow mode. It should be able to read the LVR of each attached i2c device to determine if it should be mixed fast or mixed slow (or pure if no i2c devices on the bus(

ExaltZephyr replied:
I have added the LVR logic check, but I am still verifying whether the i2c-scl-hz property is specified in the DTS, as it needs to be checked first according to the documentation in i3c-controller.yaml

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+	struct i3c_stm32_msg *curr_msg = &data->curr_msg;
+
+	/* Either should be NULL */
+	if ((i3c_msgs == NULL && i2c_msgs == NULL) || (i3c_msgs != NULL && i2c_msgs != NULL)) {
+		return -EINVAL;
+	}
```
These look like programming errors and could rather be asserts

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+static int i3c_stm32_i2c_configure(const struct device *dev, uint32_t config)
+{
+	switch (I2C_SPEED_GET(config)) {
+	case I2C_SPEED_FAST:
+	case I2C_SPEED_FAST_PLUS:
+		break;
```
This should be setting the `data->drv_data.ctrl_config.scl.i2c`

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+		*ccc_target_payload_sf;          /* Current target addressed by 2nd part of direct CCC command used by the status FIFO */
+	size_t ccc_target_idx;          /* Current target index, used for filling C-FIFO */
+	struct k_sem device_sync_sem;   /* Sync between device communication messages */
+	struct k_sem bus_mutex;         /* Sync between transfers */
```
this should be a mutex to prevent priority inversions rather than a semaphore

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+	I3C_TypeDef *i3c = config->i3c;
+	int ret;
+
+	if (addr > 0x7f) {
```
There is a 10b messaging flag for i2c it should be checking instead

XenuIsWatching reviewed (changes requested):
`i3c_shell.c` also needs updated https://github.com/zephyrproject-rtos/zephyr/blob/c50777a8431157bd4ab3a2c2e2a37e8151ae07da/drivers/i3c/i3c_shell.c#L82
and
https://github.com/zephyrproject-rtos/zephyr/blob/c50777a8431157bd4ab3a2c2e2a37e8151ae07da/drivers/i3c/i3c_shell.c#L96

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+	}
+
+	if (((target != NULL) && i3c_device_is_ibi_capable(target)) ||
+	    (((target->bcr & I3C_BCR_IBI_REQUEST_CAPABLE) == I3C_BCR_IBI_REQUEST_CAPABLE))) {
```
Isn't this the same as `i3c_device_is_ibi_capable`

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2434 @@
+}
+
+/* Handles the controller private read/write transfers */
+static int i3c_stm32_i3c_transfer(const struct device *dev, struct i3c_device_desc *target,
+				  struct i3c_msg *msgs, uint8_t num_msgs)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int ret;
+
+	/* Verify all messages */
+	for (size_t i = 0; i < num_msgs; i++) {
+		if (msgs[i].buf == NULL) {
+			return -EINVAL;
+		}
+	}
+
```
This is also be returning -ENOTSUP if any of the messages are HDR

JarmouniA reviewed (commented):
Merge commit should be dropped.

JarmouniA reviewed (changes requested):
The title of each commit should reflect the path of where changes are introduced, see the main branch for examples.

erwango commented:
About Sign-off (`Signed-off-by: EXALT Technologies <zephyr@exalt.ps>`), [the names of groups are not allowed](https://docs.zephyrproject.org/latest/contribute/guidelines.html#dco-sign-off). Please provide one or several developer name. You could use :
```
Signed-off-by: .....
Co-authored-by: .....
```

erwango commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,41 @@
+		help
+			Enable support for I3C on STM32 microcontrollers.
+
+config I3C_STM32_POLL
+		bool "I3C polling mode for STM32"
+		depends on I3C_STM32
+		default n
+		help
+			Enables polling mode for I3C on STM32 microcontrollers.
```
Question to maintainers (@dcpleung, @XenuIsWatching)
Is there a real interest in having this POLL option ?

If it's disabled by default, and doesn't provide a significant interest, it might be better to remove it right now to simplify driver code and long term maintenance.
We can find poll based code in few other drivers, but that's mostly for historical reasons as in general they are not used and present just because they where the initial method implemented.

@ExaltZephyr Any reason you'd like to keep this `I3C_STM32_POLL`option ?

dcpleung replied:
Unless there is a legitimate use case, I recommend removing it.

erwango commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,41 @@
+		default y
+		help
+			Enable support for I3C on STM32 microcontrollers.
+
```
Add `if I3C_STM32` here and remove `depends on I3C_STM32` in below symbols

erwango commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2447 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <stm32_ll_i3c.h>
+#include <stm32_ll_bus.h>
+#include <stm32_ll_rcc.h>
+#include <stm32h5xx_ll_system.h>
+#include <stm32h5xx_ll_cortex.h>
```
```suggestion
#include <stm32_ll_system.h>
#include <stm32_ll_cortex.h>
```

erwango replied:
Nit: Suggest to move stm32 includes out of `#include <zephyr...` section

erwango commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2447 @@
+
+#define DT_DRV_COMPAT st_stm32_i3c
+
+#define CONFIG_I3C_INIT_PRIORITY 50
```
Delete this and make use of already available `CONFIG_I3C_CONTROLLER_INIT_PRIORITY`

erwango commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2447 @@
+#define STM32_I3C_TRANSFER_TIMEOUT K_MSEC(100)
+
+/* Integer divison with ceiling */
+#define INT_DIV_CEIL(a, b) (((a) + (b - 1)) / b)
```
Could be replaced by equivalent `DIV_ROUND_UP` from util.h ?

erwango commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2447 @@
+	static struct i3c_stm32_data i3c_stm32_data_##index = {                                    \
+		.drv_data.ctrl_config.scl.i2c = DT_INST_PROP_OR(index, i2c_scl_hz, 0),             \
+		.drv_data.ctrl_config.scl.i3c = DT_INST_PROP_OR(index, i3c_scl_hz, 0),             \
+		STM32_I3C_DMA_CHANNEL(index, rx, RX, PERIPHERAL, MEMORY)                           \
+			STM32_I3C_DMA_CHANNEL(index, tx, TX, MEMORY, PERIPHERAL)                   \
+				STM32_I3C_DMA_CHANNEL(index, tc, TC, MEMORY, PERIPHERAL)           \
+					STM32_I3C_DMA_CHANNEL(index, rs, RS, PERIPHERAL, MEMORY)}; \
```
Nit: Indentation:
```suggestion
		STM32_I3C_DMA_CHANNEL(index, rx, RX, PERIPHERAL, MEMORY)                           \
		STM32_I3C_DMA_CHANNEL(index, tx, TX, MEMORY, PERIPHERAL)     		              \
		STM32_I3C_DMA_CHANNEL(index, tc, TC, MEMORY, PERIPHERAL)    			\
		STM32_I3C_DMA_CHANNEL(index, rs, RS, PERIPHERAL, MEMORY)}; 			\
```

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -456,7 +456,16 @@
+			interrupt-names = "event", "error";
+			#address-cells = <3>;
+			#size-cells = <0>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00800000>;
```
Please use new macros:
```suggestion
			clocks = <&rcc STM32_CLOCK(APB1, 17U)>;
```

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -456,7 +456,16 @@
 			interrupt-names = "event", "error";
 			status = "disabled";
 		};
-
```
Nit: Break a line

erwango commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -456,7 +456,16 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00800000>;
+			status = "disabled";
+		};
 		spi1: spi@40013000 {
```
Nit: Break a line

erwango commented on dts/bindings/i3c/st,stm32-i3c.yaml:
```diff
@@ -0,0 +1,47 @@
+      - 120: slot number (request which could be given by the GPDMA)
+      - STM32_DMA_PERIPH_RX: channel configuration (only for srce/dest data size, priority)
+
+      For example dmas for TX/RX/TC/RS on I3C
+          dmas = <&gpdma1 0 120 STM32_DMA_PERIPH_RX
+          &gpdma1 1 121 STM32_DMA_PERIPH_TX
+          &gpdma1 2 122 (STM32_DMA_PERIPH_TX | STM32_DMA_MEM_32BITS | STM32_DMA_PERIPH_32BITS)
+          &gpdma1 3 123 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_32BITS | STM32_DMA_PERIPH_32BITS)>;
```
Duplicate section ?

erwango commented on dts/bindings/i3c/st,stm32-i3c.yaml:
```diff
@@ -0,0 +1,47 @@
+          &gpdma1 2 122 (STM32_DMA_PERIPH_TX | STM32_DMA_MEM_32BITS | STM32_DMA_PERIPH_32BITS)
+          &gpdma1 3 123 (STM32_DMA_PERIPH_RX | STM32_DMA_MEM_32BITS | STM32_DMA_PERIPH_32BITS)>;
+
+      With, in each cell of the dmas specifier:
+      - &gpdma1: dma controller phandle
+      - 0: channel number (0 to Max-Channel minus 1). From 0 to 7 on stm32h5x.
+      - 120: slot number (request which could be given by the GPDMA)
+      - STM32_DMA_PERIPH_RX: channel configuration (only for srce/dest data size, priority)
```
Technically this should rather be descried in dma bindings, and it is already, so I 'd suggest to drop.
One of the reason is that this depends on DMA compatible version, which may vary.

erwango commented on boards/st/nucleo_h563zi/nucleo_h563zi-common.dtsi:
```diff
@@ -78,6 +78,14 @@
 	status = "okay";
 };
 
+&i3c1 {
+	pinctrl-0 =  <&i3c1_scl_pd12 &i3c1_sda_pd13>;
+	pinctrl-names = "default";
+	i3c-scl-hz = <12500000>;
+	i2c-scl-hz = <400000>;
+	status = "okay";
+};
```
Please also update board documentation

erwango commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,41 @@
+		help
+			Enable support for I3C on STM32 microcontrollers.
+
+config I3C_STM32_POLL
```
Let's drop it, as suggested by maintainer.

erwango reviewed (changes requested):
Some comments but nice work for a first contribution!

erwango commented on drivers/i3c/i3c_shell.c:
```diff
@@ -94,8 +95,9 @@ DT_FOREACH_STATUS_OKAY(nxp_mcux_i3c, I3C_CTRL_FN)
 	/* zephyr-keep-sorted-start */
 	DT_FOREACH_STATUS_OKAY(cdns_i3c, I3C_CTRL_LIST_ENTRY)
-	DT_FOREACH_STATUS_OKAY(nuvoton_npcx_i3c, I3C_CTRL_LIST_ENTRY)
-	DT_FOREACH_STATUS_OKAY(nxp_mcux_i3c, I3C_CTRL_LIST_ENTRY)
+		DT_FOREACH_STATUS_OKAY(st_stm32_i3c, I3C_CTRL_LIST_ENTRY)
+			DT_FOREACH_STATUS_OKAY(nuvoton_npcx_i3c, I3C_CTRL_LIST_ENTRY)
+				DT_FOREACH_STATUS_OKAY(nxp_mcux_i3c, I3C_CTRL_LIST_ENTRY)
```
Why changing indentation ?

ExaltZephyr replied:
Hi @erwango,
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11894800301/job/33149043014?pr=81190
This job failed due to a formatting issue. I’ll revert the formatting changes and address them in a separate commit to resolve this. ok ?

erwango commented on drivers/i3c/i3c_shell.c:
```diff
@@ -2104,29 +2105,28 @@ SHELL_DYNAMIC_CMD_CREATE(dsub_i3c_device_name, i3c_device_name_get);
-		      "Usage: ibi disable <device> <target>",
-		      cmd_i3c_ibi_disable, 3, 0),
-	SHELL_SUBCMD_SET_END /* Array terminated. */
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_i3c_ibi_cmds,
+			       SHELL_CMD_ARG(hj, &dsub_i3c_device_name,
+					     "Send IBI HJ\n"
+					     "Usage: ibi hj <device>",
+					     cmd_i3c_ibi_hj, 2, 0),
+			       SHELL_CMD_ARG(tir, &dsub_i3c_device_name,
+					     "Send IBI TIR\n"
+					     "Usage: ibi tir <device> [<byte1>, ...]",
+					     cmd_i3c_ibi_tir, 2, MAX_I3C_BYTES),
+			       SHELL_CMD_ARG(cr, &dsub_i3c_device_name,
+					     "Send IBI CR\n"
+					     "Usage: ibi cr <device>",
+					     cmd_i3c_ibi_cr, 2, 0),
+			       SHELL_CMD_ARG(enable, &dsub_i3c_device_attached_name,
+					     "Enable receiving IBI from target\n"
+					     "Usage: ibi enable <device> <target>",
+					     cmd_i3c_ibi_enable, 3, 0),
+			       SHELL_CMD_ARG(disable, &dsub_i3c_device_attached_name,
+					     "Disable receiving IBI from target\n"
+					     "Usage: ibi disable <device> <target>",
+					     cmd_i3c_ibi_disable, 3, 0),
+			       SHELL_SUBCMD_SET_END /* Array terminated. */
 );
```
Don't mix formatting changes with other changes. 
Please do it in a dedicated commit if required.

erwango commented on drivers/i3c/i3c_shell.c:
```diff
@@ -2145,13 +2145,12 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
-		      "Usage: hdr ddr <sub cmd>",
-		      NULL, 2, 0),
-	SHELL_SUBCMD_SET_END /* Array terminated. */
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_i3c_hdr_cmds,
+			       SHELL_CMD_ARG(ddr, &sub_i3c_hdr_ddr_cmds,
+					     "Send HDR DDR\n"
+					     "Usage: hdr ddr <sub cmd>",
+					     NULL, 2, 0),
+			       SHELL_SUBCMD_SET_END /* Array terminated. */
```
Ditto

erwango commented on boards/st/nucleo_h563zi/doc/index.rst:
```diff
@@ -1,4 +1,7 @@
-.. zephyr:board:: nucleo_h563zi
+.. _nucleo_h563zi_board:
+
+ST Nucleo H563ZI
+################
```
Please revert to original code (see upstream changes)

erwango commented on boards/st/nucleo_h563zi/doc/index.rst:
```diff
@@ -38,6 +41,10 @@ Here are some highlights of the Nucleo H563ZI board:
 
 More information about the board can be found at the `NUCLEO_H563ZI website`_.
 
+.. image:: img/nucleo_h563zi.jpg
+   :align: center
+   :alt: NUCLEO H563ZI
+
```
Please revert to original code (see upstream changes)

erwango reviewed (changes requested):
Thanks for the changes. Few more comments, but for my side it looks close to completion.

Minor issue in commit message:
```diff
-Co-authored-by: Sara Touqan <zephyr@exalt.pst status
->
+Co-authored-by: Sara Touqan <zephyr@exalt.pst>
```

Also, as requested by @JarmouniA, please split into several commits:
- Dts changes
- Driver introduction
- Shell update
- Commit for formatting changes if needed
- Board enablement

erwango commented:
Please also fix https://github.com/zephyrproject-rtos/zephyr/actions/runs/11893617721/job/33140221110?pr=81190#step:11:45

ExaltZephyr commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,36 @@
+config I3C_STM32
+	bool "STM32 I3C driver support"
+	depends on DT_HAS_ST_STM32_I3C_ENABLED
+	select USE_STM32_HAL_I3C
```
Initially, we attempted to use USE_STM32_LL_I3C, but encountered compilation issues. 
As a workaround, we switched to USE_STM32_HAL_I3C to make LL work.

erwango replied:
Ok, anyway, you should be able to remove now.

And, out of curiosity, why didn't you use HAL API instead of LL ?

erwango commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,36 @@
+	bool "STM32 I3C driver support"
+	depends on DT_HAS_ST_STM32_I3C_ENABLED
+	select USE_STM32_HAL_I3C
+	default y
```
Looks like you miss a `select I3C_IBI_WORKQUEUE if I3C_USE_IBI`, if we look at other drivers, otherwise it doesn't compile.

This being said, I wonder why `I3C_IBI_WORKQUEUE` isn't defined as `default y`. @XenuIsWatching maybe you'll be able to tell ?

erwango commented on drivers/i3c/Kconfig.stm32:
```diff
@@ -0,0 +1,36 @@
+config I3C_STM32
+	bool "STM32 I3C driver support"
+	depends on DT_HAS_ST_STM32_I3C_ENABLED
+	select USE_STM32_HAL_I3C
```
I confirm this is not required. Please remove

ExaltZephyr replied:
@erwango  We chose to use LL instead of HAL, as we took i2c_stm32 as a reference and assumed that following the LL convention would align better with the project's standards.

erwango reviewed (changes requested):
Last remaining issue on my side is the fact that we need split commit as mentioned above. Also, I still need to test this, hence the "Block HW test" label, I'll remove as soon as we've been able to test.

erwango commented:
@XenuIsWatching PTAL

ExaltZephyr committed:
dts: Add I3C configuration for STM32.

This commit adds the main DTS configurations required
to enable I3C support on STM32.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: i3c: Support I3C driver for STM32.

This commit introduces support for the I3C driver on STM32, enabling
functionality APIs for I3C controllers.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr committed:
shell: Enable I3C shell support for STM32.

This commit introduces support for I3C shell on STM32.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr committed:
boards: Enable I3C support for STM32 nucleo_h563zi.

This commit enables I3C support for STM32 nucleo_h563zi boards.

Signed-off-by: Mohammad Badawi <zephyr@exalt.ps>
Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr commented:
done

ExaltZephyr commented on dts/bindings/test/vnd,i3c.yaml:
```diff
@@ -15,3 +15,9 @@ include:
 properties:
   label:
     type: string
+
+  supports-setaasa:
+    type: boolean
+    description: |
+      Indicates if the device supports the CCC SETAASA. If true, it will
+      be used as an optimization for bus initialization.
```
@erwango 
The reason for adding this change was an issue I encountered during compilation when defining I3C targets in the overlay file for the samples. The error in the DTS reported something like :
' error: 'DT_N_S_soc_S_i3c_40005c00_S_i3cdev0_000000020813811000_P_supports_setaasa' undeclared here...'

Upon investigating the code, I found that the issue comming from the following macro in Zephyr:

```c #define I3C_DEVICE_DESC_DT(node_id)                \
    {                                               \
        .bus = DEVICE_DT_GET(DT_BUS(node_id)),      \
        .dev = DEVICE_DT_GET(node_id),              \
        .static_addr = DT_PROP_BY_IDX(node_id, reg, 0), \
        .pid = ((uint64_t)DT_PROP_BY_IDX(node_id, reg, 1) << 32) \
            | DT_PROP_BY_IDX(node_id, reg, 2),     \
        .init_dynamic_addr =                        \
            DT_PROP_OR(node_id, assigned_address, 0), \
        .supports_setaasa = DT_PROP(node_id, supports_setaasa), \
    }, ```


The problem happens because DT_PROP is used to directly fetch the supports_setaasa property, but if the property is missing, the macro does not handle it.

To fix the issue, we have two possible approaches:

1) Add the supports_setaasa property in the relevant yaml, as I did.
2) Propose a change in Zephyr to use DT_PROP_OR instead of DT_PROP. This way, the macro can provide a default value (e.g., 0) when the property is not explicitly defined in the device tree. 

which approach do you prefer ?

erwango replied:
Ok, hence the need for documenting commit messages ;).
I'd opt for option 3: Add this property in `i3c-controller.yaml`

ExaltZephyr replied:
Yes, I can add that in a separate commit.
Just to clarify, the issue was raised for the i3c targets, not the controllers. The previous macro will be called when looping over the targets.
Since I used compatible: "vnd,i3c" for the targets, I have added the property in the vnd,i3c.yaml file

XenuIsWatching commented on dts/bindings/i3c/i3c-controller.yaml:
```diff
@@ -30,3 +30,9 @@ properties:
       Virtual Register (LVR) of all connected I2C devices to determine
       the maximum allowed frequency.
+
+  supports-setaasa:
+    type: boolean
+    description: |
+      Indicates if the device supports the CCC SETAASA. If true, it will
+      be used as an optimization for bus initialization.
```
This shouldn't be added here, remove

ExaltZephyr replied:
The reason for adding this change was an issue I encountered during compilation when adding I3C targets in the overlay file for the samples. The error in the DTS reported something like :
' error: 'DT_N_S_soc_S_i3c_40005c00_S_i3cdev0_000000020813811000_P_supports_setaasa' undeclared here...'

Upon investigating the code, I found that the issue comming from the following macro in Zephyr:

#define I3C_DEVICE_DESC_DT(node_id)
{
.bus = DEVICE_DT_GET(DT_BUS(node_id)),
.dev = DEVICE_DT_GET(node_id),
.static_addr = DT_PROP_BY_IDX(node_id, reg, 0),
.pid = ((uint64_t)DT_PROP_BY_IDX(node_id, reg, 1) << 32)
| DT_PROP_BY_IDX(node_id, reg, 2),
.init_dynamic_addr =
DT_PROP_OR(node_id, assigned_address, 0),
.supports_setaasa = DT_PROP(node_id, supports_setaasa),
}

The problem happens because DT_PROP is used to directly fetch the supports_setaasa property, but if the property is missing, the macro does not handle it.

XenuIsWatching replied:
What is the overlay here? This really looks like you are using the wrong the include with the dts bindings for what you have in your overlay?

XenuIsWatching replied:
@ExaltZephyr I still need a response to this

ExaltZephyr replied:
Overlays in Zephyr are files used within samples to extend or modify the DTS configurations for a board. The compilation error occurred because a required property was missing in the I3C YAML files. I added the missing property to ensure that future sample additions won't face the same issue that I faced when I added my sample to test
ok ?

XenuIsWatching replied:
no, not okay, I know what overlays are, and I've seen people internally see the same error that you are seeing, and it's always because of having a wrong binding in their yaml file. What sample are you referring too? I don't see anything in this PR.

There should be `include: ["i3c-device.yaml"]` within there not `include: ["i3c-controller.yaml"]`. This commit must be dropped

ExaltZephyr replied:
The sample I'm referring to is a new one that I added locally just for testing.

Anyway, I will drop the commit now

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+						"frequency for FM I2C bus");
+					return -EINVAL;
+				}
+			} else {
```
does this not support MIXED_SLOW?

ExaltZephyr replied:
yes this does not support MIXED_SLOW

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	LL_I3C_EnableStatusFIFO(i3c);
+
+	/* Configure Controller */
+	LL_I3C_SetOwnDynamicAddress(i3c, 0);
```
There shouldn't be a dynamic address of 0. use one of the i3c address apis to pick one (or assign one with a dts prop) if it is a primary controller

XenuIsWatching replied:
So... this was just removed? I'm okay with removing it as target mode (nor handoff) is supported by this driver yet.

ExaltZephyr replied:
It was never intended for the controller to have a dynamic address to begin with, that's why we decided to remove it

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	int ret;
+
+	if (type == I3C_CONFIG_TARGET || type == I3C_CONFIG_CUSTOM) {
+		return -EOPNOTSUPP;
```
`-ENOTSUP`

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	}
+
+	/* I3C Initialization */
+	LL_I3C_SetMode(i3c, LL_I3C_MODE_CONTROLLER);
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
+	LL_I3C_SetControllerActivityState(i3c, LL_I3C_OWN_ACTIVITY_STATE_0);
```
I would expect these to be in the `init` function, not in the configure function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	/* Configure FIFO */
+	LL_I3C_SetRxFIFOThreshold(i3c, LL_I3C_RXFIFO_THRESHOLD_1_4);
+	LL_I3C_SetTxFIFOThreshold(i3c, LL_I3C_TXFIFO_THRESHOLD_1_4);
+	LL_I3C_EnableControlFIFO(i3c);
+	LL_I3C_EnableStatusFIFO(i3c);
+
+	/* Configure Controller */
+	LL_I3C_SetOwnDynamicAddress(i3c, 0);
+	LL_I3C_EnableOwnDynAddress(i3c);
+	LL_I3C_SetStallTime(i3c, 0x00);
+	LL_I3C_DisableStallACK(i3c);
+	LL_I3C_DisableStallParityCCC(i3c);
+	LL_I3C_DisableStallParityData(i3c);
+	LL_I3C_DisableStallTbit(i3c);
+	LL_I3C_DisableHighKeeperSDA(i3c);
+
+	LL_I3C_Enable(i3c);
+
+	LL_I3C_EnableIT_FC(i3c);
+	LL_I3C_EnableIT_CFNF(i3c);
+	LL_I3C_EnableIT_SFNE(i3c);
+	LL_I3C_EnableIT_RXFNE(i3c);
+	LL_I3C_EnableIT_TXFNF(i3c);
+	LL_I3C_EnableIT_ERR(i3c);
+	LL_I3C_EnableIT_WKP(i3c);
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableIT_IBI(i3c);
+	LL_I3C_EnableIT_HJ(i3c);
+#endif
+
+	/* Bus will be idle initially */
+	data->msg_state = STM32_I3C_MSG_IDLE;
+	data->sf_state = STM32_I3C_SF_IDLE;
+	data->target_id = 0;
+#ifdef CONFIG_I3C_USE_IBI
+	data->ibi_payload = 0;
+	data->ibi_payload_size = 0;
+	data->ibi_target_addr = 0;
+#endif
```
I would expect these to be in the `init` function, not in the configure function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	if (dev == NULL || payload == NULL) {
+		return -EINVAL;
+	}
```
this look like programing errors, I'd recommend using ASSERTS here instead

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	payload->ccc.num_xfer = 0;
+
+	for (size_t i = 0; i < payload->targets.num_targets; i++) {
+		payload->targets.payloads->num_xfer = 0;
```
this is only setting the first payload over and and over again with 0. I would expect this
```c
payload->targets.payloads[i].num_xfer = 0;
```

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+			LL_I3C_TransmitData8(i3c, target->data[target->num_xfer++]);
+
+			/* After sending all bytes for current target, move on to the next target */
+			if (target->num_xfer == target->data_len) {
```
num_xfer can sometimes not equal the `data_len` because in I3C a target can send a EoD before the expected `data_len` such as with GETCAPS and GETMXDS which can have a variable length. Does this handle this situation?

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+
+	aval = DIV_ROUND_UP(1000ull * i3c_clock, 1000000000ull) - 1;
+
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
```
this appears to be set in multiple places, it probably should be only set once in the init

XenuIsWatching replied:
not resolved!!

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
+	LL_I3C_SetFreeTiming(i3c, free_timing);
+	LL_I3C_SetControllerActivityState(i3c, LL_I3C_OWN_ACTIVITY_STATE_0);
```
this appears to be set in multiple places, it probably should be only set once in the init

XenuIsWatching replied:
not resolved!!

ExaltZephyr replied:
In this section, I calculate the free_timing value based on specific calculations and use it with LL_I3C_SetFreeTiming. The functions LL_I3C_SetDataHoldTime and LL_I3C_SetControllerActivityState are included alongside LL_I3C_SetFreeTiming as they all modify the same register.
ok ?

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+}
+
+/* Configures the I3C module in controller mode */
+static int i3c_stm32_configure(const struct device *dev, enum i3c_config_type type, void *cfg)
```
where is `cfg` used in this function? I would expect a way to configure the i3c/i2c speed at runtime with in here to configure the values with in `data->common.ctrl_config`

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_get(dev);
+#endif
+	/* Prevent the clocks to be stopped during the transaction */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	/* Mark current transfer as DAA */
+	data->msg_state = STM32_I3C_MSG_DAA;
+
+	/* Disable TXFNF interrupt, the RXFNE interrupt will enable it once all PID bytes are
+	 * received
+	 */
+	LL_I3C_DisableIT_TXFNF(i3c);
+
+	/* Start DAA */
+	LL_I3C_ControllerHandleCCC(i3c, I3C_CCC_ENTDAA, 0, LL_I3C_GENERATE_STOP);
+
+	/* Wait for DAA to finish */
+	if (k_sem_take(&data->device_sync_sem, STM32_I3C_TRANSFER_TIMEOUT) != 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (data->msg_state == STM32_I3C_MSG_ERR) {
+		i3c_stm32_clear_err(dev, false);
+		/* Enable TXFNF interrupt in case an error occurred before it was enabled by RXFNE
+		 */
+		LL_I3C_EnableIT_TXFNF(i3c);
+		return -EIO;
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
```
How is this to gather what it found with ENTDAA and the values it assigned?

ExaltZephyr replied:
Are you asking how to retrieve data like the PID from the target? If so, you can refer to the logic implemented in the ISR, particularly in the function i3c_stm32_event_isr_rx.

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+	for (size_t i = 0; i < num_msgs; i++) {
+		if (msgs[i].buf == NULL) {
+			return -EINVAL;
+		}
+		if ((msgs[i].flags & I3C_MSG_HDR) && (msgs[i].hdr_mode != 0)) {
+			return -ENOTSUP;
+		}
+	}
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+	ret = i3c_stm32_curr_msg_init(dev, msgs, NULL, num_msgs, target->dynamic_addr);
+	if (ret != 0) {
+		i3c_stm32_clear_err(dev, false);
+		LOG_ERR("Failed to initialize transfer messages, err=%d", ret);
+		return ret;
+	}
+
+	ret = i3c_stm32_transfer_begin(dev);
+	if (ret != 0) {
+		i3c_stm32_clear_err(dev, false);
+		LOG_ERR("Failed to transfer messages, err=%d", ret);
+		return ret;
+	}
+
+#ifdef CONFIG_I3C_STM32_DMA
+	/* Fill the num_xfer for each message from the status FIFO */
+	for (size_t i = 0; i < num_msgs; i++) {
+		msgs[i].num_xfer = READ_BIT(data->status_fifo[i], I3C_SR_XDCNT);
+	}
+
+	k_heap_free(&stm32_i3c_fifo_heap, data->control_fifo);
+	k_heap_free(&stm32_i3c_fifo_heap, data->status_fifo);
+
+	i3c_stm32_end_dma_requests(dev);
+#endif
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
```
There is the `I3C_MSG_NBCH` within `flags` that should be used to disable/enable the arbitration 0x7e flag

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2129 @@
+
+			/* After receiving all bytes for current target, move on to the next target
+			 */
+			if (target->num_xfer == target->data_len) {
```
A target can send an EoD when doing a read with a CCC and it's possible for the num_xfer to be less than the data_len. This can happen with variable length CCCs such as GETCAPS and GETMXDS. Does this handle this situation?

ExaltZephyr replied:
No it is not handled, we are currently thinking of a way to handle this issue. The problem is that the information about the target early read termination can be obtained from the status FIFO, however, we might be receiving data from the next target before the data from the status FIFO is ready.

One suggestion we are thinking of is to disable the status FIFO for CCC requests and rely on the RXTGTENDF interrupt to detect early target termination and advance the target_ccc_payload pointer.

XenuIsWatching replied:
do you plan to address this as apart of this PR?

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+			LL_I3C_EnableArbitrationHeader(i3c);
+		}
+	} else {
+		if (curr_msg->i2c_msg_ctrl_ptr->flags & I3C_MSG_NBCH) {
```
i2c does not have this flag

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+	struct i3c_stm32_data *data = dev->data;
+	struct i3c_config_controller *ctrl_config = &data->drv_data.ctrl_config;
+
+	switch (I2C_SPEED_GET(config)) {
+	case I2C_SPEED_FAST:
+		ctrl_config->scl.i2c = 400000;
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		ctrl_config->scl.i2c = 1000000;
+		break;
+	default:
+		return -EINVAL;
+	}
```
Are other speeds not supported such as `I2C_SPEED_STANDARD`?

ExaltZephyr replied:
The only 2 I2C speeds supported for this chip are the FM and the FM+

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+	data->msg_state = STM32_I3C_MSG;
+	data->sf_state = STM32_I3C_SF;
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
```
nit: these #ifdef are not neccessary around `pm_device_runtime_*` as it will look at that kconfig within the function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+	struct i3c_stm32_data *data = dev->data;
+	struct i3c_config_controller *ctrl_cfg = cfg;
+
+	if ((ctrl_cfg != NULL) && ((ctrl_cfg->scl.i2c == 0U) || (ctrl_cfg->scl.i3c == 0U))) {
+		return -EINVAL;
+	}
+
+	if (ctrl_cfg != NULL) {
+		data->drv_data.ctrl_config.scl.i3c = ctrl_cfg->scl.i3c;
+		data->drv_data.ctrl_config.scl.i2c = ctrl_cfg->scl.i2c;
+	}
```
I can't think of a reason as to why you would be passing in NULL as `cfg` to this function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+	i3c_stm32_configure(dev, I3C_CONFIG_CONTROLLER, NULL);
+	i3c_stm32_controller_init(dev);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_enable(dev);
+#endif
```
This should not be here, this would give the user no way to disable this individually.
If this is to be called, then there is to be `zephyr,pm-device-runtime-auto` within the dts node for this and will be called right after the init function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+
+#ifdef CONFIG_I3C_USE_IBI
+
+int i3c_stm32_ibi_enable(const struct device *dev, struct i3c_device_desc *target)
```
shouldn't you be calling `pm_device_enable_get(dev)` within here

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+	return ret;
+}
+
+int i3c_stm32_ibi_disable(const struct device *dev, struct i3c_device_desc *target)
```
shouldn't you be calling `pm_device_enable_put(dev)` within here

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2163 @@
+		LOG_ERR("Error sending IBI ENEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	if (i3c_device_is_ibi_capable(target)) {
```
above you check if it is not ibi capable and return an error code. so this statement will always ever be `true`. do you mean to check if it has a payload with an ibi especially for that `IBIAddData` argument with `i3c_ibi_has_payload`

XenuIsWatching commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -457,6 +457,17 @@
 			status = "disabled";
 		};
 
+		i3c1: i3c@40005c00 {
+			compatible = "st,stm32-i3c";
+			reg = <0x40005c00 0x1000>;
+			interrupts = <123 0>, <124 0>;
+			interrupt-names = "event", "error";
+			#address-cells = <3>;
+			#size-cells = <0>;
+			clocks = <&rcc STM32_CLOCK(APB1, 23U)>;
+			status = "disabled";
+		};
```
There is also i3c2
```dts
		i3c1: i3c@44003000 {
			compatible = "st,stm32-i3c";
			reg = <0x44003000 0x400>;
			interrupts = <131 0>, <132 0>;
			interrupt-names = "event", "error";
			#address-cells = <3>;
			#size-cells = <0>;
			clocks = <&rcc STM32_CLOCK(APB3, 9U)>;
			status = "disabled";
		};
```

XenuIsWatching commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -457,6 +457,17 @@
 
+		i3c1: i3c@40005c00 {
+			compatible = "st,stm32-i3c";
+			reg = <0x40005c00 0x1000>;
```
register space is 0x400
```
reg = <0x40005c00 0x400>;
```

XenuIsWatching commented on dts/arm/st/h5/stm32h5.dtsi:
```diff
@@ -457,6 +457,32 @@
+		i3c2: i3c@44003000 {
+			compatible = "st,stm32-i3c";
+			reg = <0x44003000 0x400>;
+			interrupts = <123 0>, <124 0>;
```
these interrupts are not correct

dcpleung commented:
Please do not resolve comments on your own, and let the original poster do it instead.

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2149 @@
+	I3C_TypeDef *i3c = config->i3c;
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableHJAck(i3c);
```
you may want to enable HJ Acks after the `i3c_bus_init` in the init function, otherwise there could be device request an assignment before the bus is ready

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2149 @@
+			return ret;
+		}
+	}
+
```
Enabling HJ should be here

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2149 @@
+		LOG_ERR("Error sending IBI DISEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	if (((target != NULL) && i3c_device_is_ibi_capable(target)) ||
+	    (((target->bcr & I3C_BCR_IBI_REQUEST_CAPABLE) == I3C_BCR_IBI_REQUEST_CAPABLE))) {
```
This is checking the same thing twice
you only just need to call `if ((target != NULL) && i3c_device_is_ibi_capable(target))`

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2149 @@
+
+	/* Set I3C bus devices configuration */
+	LL_I3C_ConfigDeviceCapabilities(i3c, (idx + 1), target->dynamic_addr,
+					LL_I3C_IBI_CAPABILITY, LL_I3C_IBI_DATA_ENABLE,
```
Shouldn't `LL_I3C_IBI_DATA_ENABLE` be checked with `i3c_ibi_has_payload`

XenuIsWatching replied:
This is not resolved

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2149 @@
+	LL_I3C_DisableStallParityData(i3c);
+	LL_I3C_DisableStallTbit(i3c);
+	LL_I3C_DisableHighKeeperSDA(i3c);
+	LL_I3C_SetDataHoldTime(i3c, LL_I3C_SDA_HOLD_TIME_1_5);
```
Thinking more about this, This is related to the kernel clock input and meeting the required t_hd_pp time, and should be within `i3c_stm32_config_ctrl_bus_char`... but I would expect some math behind it to determine if it should be 0.5 or 1.5.

ExaltZephyr replied:
According to the reference manual for the H5 board, the value of the minimum data hold time can be calculated as follows:

- $t_{HDPP_{min}} = min(t_{CR}, t_{CF}) + 3ns$

We could not find how the $t_{CR}$ and $t_{CF}$ values can be calculated so we chose 1.5 data hold time as default for now to avoid any issues that might occur when choosing 0.5.

If you have any suggestion on how we can dynamically set these values according to the $t_{CR}$, $t_{CF}$ and the kernel clock, we can modify the code accordingly.

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2171 @@
+		/* Set I3C bus devices configuration */
+		LL_I3C_ConfigDeviceCapabilities(i3c, (idx + 1), target->dynamic_addr,
+						LL_I3C_IBI_NO_CAPABILITY,
+						i3c_ibi_has_payload(target) ? LL_I3C_IBI_DATA_ENABLE : LL_I3C_IBI_DATA_DISABLE,
```
This should only be LL_I3C_IBI_DATA_DISABLE, this `i3c_ibi_has_payload(target) ? LL_I3C_IBI_DATA_ENABLE : LL_I3C_IBI_DATA_DISABLE,` should only be with the `ibi_enable` function

XenuIsWatching commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2171 @@
+		LOG_ERR("Error sending IBI DISEC for 0x%02x (%d)", target->dynamic_addr, ret);
+	}
+
+	if ((target != NULL) && i3c_device_is_ibi_capable(target)) {
```
The check for `i3c_device_is_ibi_capable` is unnecessary here as it is already checking for it above and returning

pdgendt commented on drivers/i3c/i3c_stm32.c:
```diff
@@ -0,0 +1,2171 @@
+
+#endif
+
+static const struct i3c_driver_api i3c_stm32_driver_api = {
```
Device driver APIs need to use the `DEVICE_API` wrapper.

```suggestion
static DEVICE_API(i3c, i3c_stm32_driver_api) = {
```

erwango replied:
@ExaltZephyr Note that this is a new macro and this may require a rebase.

XenuIsWatching reviewed (dismissed):
LGTM now... but it looks like there is an unrelated CI failure, you may need to rebase back on top of main

erwango commented:
Thanks @XenuIsWatching and @ExaltZephyr for the thorough review process! 

> LGTM now... but it looks like there is an unrelated CI failure, you may need to rebase back on top of main

Need to wait https://github.com/zephyrproject-rtos/zephyr/pull/82363 to be merged.

XenuIsWatching commented on dts/bindings/test/vnd,i3c.yaml:
```diff
@@ -9,3 +9,4 @@ compatible: "vnd,i3c"
 
 include:
   - name: i3c-controller.yaml
+  - name: i3c-device.yaml
```
remove this commit...
you are still using the wrong compat, `vnd,i3c-device` has what you should be using

ExaltZephyr replied:
I included this file because the "supports-setaasa" property is required for my i3c samples.

ExaltZephyr replied:
When I used the compat "vnd,i3c-device" under samples, I encountered a compilation error. The reason for this error was that we do not have the creation of the device object from the device tree node identifier for this compat.

Using compat as  "vnd,i3c" worked with me under samples because we already have the creation of the device object for this compatibility under the drivers/i3c_test file.

XenuIsWatching replied:
Remove this commit again, Please see my earlier comments... you are still doing the same issue I mentioned earlier

XenuIsWatching reviewed (changes requested):
see comment

erwango commented:
Please rebase to fix twister CI issues (fixed by https://github.com/zephyrproject-rtos/zephyr/pull/82497)

ExaltZephyr commented:
The scancode job has been stuck since yesterday. 
Do you have any suggestions on how to resolve this issue? 
I noticed that I do not have the permissions to retrigger it

kartben commented:
> The scancode job has been stuck since yesterday. Do you have any suggestions on how to resolve this issue? I noticed that I do not have the permissions to retrigger it

Should be back now. GitHub glitch, it would seem.
